### PR TITLE
Use `.tail` prefix to optimize JIT-generated code

### PR DESCRIPTION
**Pull request by SergeiPavlov**
The effect is shown here:
https://github.com/servicetitan/dataobjects-net/pull/156

### CODE REVIEW

File: LazyProxy/LazyProxyBuilder.cs [hunk 1]
```diff
@@ -235,6 +235,7 @@ private static TypeBuilder AddMethods(this TypeBuilder typeBuilder, Type type, F
                     generator.Emit(OpCodes.Ldarg, i);
                 }
 
+                generator.Emit(OpCodes.Tail);
```
Comment (hypercodeplace) [inline LazyProxy/LazyProxyBuilder.cs:238]: Just wondering, why is the difference with [OpCodes.Tailcall](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.emit.opcodes.tailcall?view=net-8.0)? And also, does this optimization work for all calls or for the recursion calls only?

Comment (SergeiPavlov) [reply LazyProxy/LazyProxyBuilder.cs]: Ah, `OpCodes.Tailcall` is correct vor .NET (fixed the code)
And `.Tail`  is for `Cecil` library used in DO

It works only before sequence `call`, `ret`
And this is not always recursive

**Comment by botinko** [comment]
I cannot estimate how much this change may affect the reliability of the application. As noted in the article, this optimization is not used by default. This means that the behavior of the JIT compiler may not be sufficiently tested. Also I afraid of increased JIT time because of optimization.

**Review by hypercodeplace — APPROVED** [review]

**Comment by SergeiPavlov** [comment]
> I cannot estimate how much this change may affect the reliability of the application. As noted in the article, this optimization is not used by default. This means that the behavior of the JIT compiler may not be sufficiently tested. Also I afraid of increased JIT time because of optimization.

`.tail` prefix tells JIT just to convert `call, ret` intio `jmp` without additional code analyzis.
It means less work for JIT.

Looks like no potential flaw. This prefix is actively used by functional-language compilers like `F#`, where tail recursion is critical feature.

**Review by botinko — APPROVED** [review]

**Pull request merged**